src/agent/src/netlink2.rs

```diff
@@ -0,0 +1,1020 @@
+            let routes4 = self
+                .handle
+                .route()
+                .get(IpVersion::V4)
+                .execute()
+                .try_collect::<Vec<_>>()
+                .await
+                .with_context(|| "Failed to query IP v4 routes")?;
+
+            let routes6 = self
+                .handle
+                .route()
+                .get(IpVersion::V6)
+                .execute()
+                .try_collect::<Vec<_>>()
+                .await
+                .with_context(|| "Failed to query IP v6 routes")?;
+
+            [routes4, routes6].concat()
+        };
+
+        Ok(list)
+    }
+
+    pub async fn list_routes(&self) -> Result<Vec<Route>> {
+        let mut result = Vec::new();
+
+        for msg in self.query_routes(None).await? {
+            // Ignore non-main tables
+            if msg.header.table != packet::constants::RT_TABLE_MAIN {
+                continue;
+            }
+
+            let mut route = Route {
+                scope: msg.header.scope as _,
+                ..Default::default()
+            };
+
+            if let Some((ip, mask)) = msg.destination_prefix() {
+                route.dest = format!("{}/{}", ip, mask);
+            }
+
+            if let Some((ip, mask)) = msg.source_prefix() {
+                route.source = format!("{}/{}", ip, mask);
+            }
+
+            if let Some(addr) = msg.gateway() {
+                route.gateway = addr.to_string();
+
+                // For gateway, destination is 0.0.0.0
+                route.dest = if addr.is_ipv4() {
+                    String::from("0.0.0.0")
+                } else {
+                    String::from("::1")
+                }
+            }
+
+            if let Some(index) = msg.output_interface() {
+                route.device = self.find_link(LinkFilter::Index(index)).await?.name();
+            }
+
+            result.push(route);
+        }
+
+        Ok(result)
+    }
+
+    /// Adds a list of routes from iterable object `I`.
+    /// It can accept both a collection of routes or a single item (via `iter::once()`).
+    /// It'll also take care of proper order when adding routes (gateways first, everything else after).
+    async fn add_routes<I>(&mut self, list: I) -> Result<()>
+    where
+        I: IntoIterator<Item = Route>,
+    {
+        // Split the list so we add routes with no gateway first.
+        // Note: `partition_in_place` is a better fit here, since it reorders things inplace (instead of
+        // allocating two separate collections), however it's not yet in stable Rust.
+        let (a, b): (Vec<Route>, Vec<Route>) = list.into_iter().partition(|p| p.gateway.len() == 0);
+        let list = a.iter().chain(&b);
+
+        for route in list {
+            let link = self.find_link(LinkFilter::Name(&route.device)).await?;
+            let is_v6 = route.gateway.contains("::") || route.dest.contains("::");
```

Comment:
We might need a more robust way to determine if it is ipv6. "::" indicates ipv6 address contains zeros? it might not be the case
Maybe we can try to parse ipaddr as v4and v6, to determine it is v4 or v6

Reply:
Updated